File: crates/topcoat-runtime/browser/src/surrogate/event.ts
import { Bool } from "./bool";
import { F64 } from "./f64";
import { String as RuntimeString } from "./string";

type EventTargetLike = globalThis.EventTarget | null;

export class Event {
	constructor(private readonly inner: globalThis.Event) {}

	get alt_key(): Bool {
		return boolProp(this.inner, "altKey");
	}

	get bubbles(): Bool {
		return new Bool(this.inner.bubbles);
	}

	get button(): F64 {
		return f64Prop(this.inner, "button");
	}

	get buttons(): F64 {
		return f64Prop(this.inner, "buttons");
	}

	get cancelable(): Bool {
		return new Bool(this.inner.cancelable);
	}

	get client_x(): F64 {
		return f64Prop(this.inner, "clientX");
	}

	get client_y(): F64 {
		return f64Prop(this.inner, "clientY");
	}

	get code(): RuntimeString {
		return stringProp(this.inner, "code");
	}

	get ctrl_key(): Bool {
		return boolProp(this.inner, "ctrlKey");
	}

	get current_target(): EventTarget {
		return new EventTarget(this.inner.currentTarget);
	}

	get data(): RuntimeString {
		return stringProp(this.inner, "data");
	}

	get default_prevented(): Bool {
		return new Bool(this.inner.defaultPrevented);
	}

	get delta_x(): F64 {
		return f64Prop(this.inner, "deltaX");
	}

	get delta_y(): F64 {
		return f64Prop(this.inner, "deltaY");
	}

	get delta_z(): F64 {
		return f64Prop(this.inner, "deltaZ");
	}

	get event_type(): RuntimeString {
		return new RuntimeString(this.inner.type);
	}

	get input_type(): RuntimeString {
		return stringProp(this.inner, "inputType");
	}

	get is_composing(): Bool {
		return boolProp(this.inner, "isComposing");
	}

	get key(): RuntimeString {
		return stringProp(this.inner, "key");
	}

	get meta_key(): Bool {
		return boolProp(this.inner, "metaKey");
	}

	get movement_x(): F64 {
		return f64Prop(this.inner, "movementX");
	}

	get movement_y(): F64 {
		return f64Prop(this.inner, "movementY");
	}

	get offset_x(): F64 {
		return f64Prop(this.inner, "offsetX");
	}

	get offset_y(): F64 {
		return f64Prop(this.inner, "offsetY");
	}

	get page_x(): F64 {
		return f64Prop(this.inner, "pageX");
	}

	get page_y(): F64 {
		return f64Prop(this.inner, "pageY");
	}

	get pointer_id(): F64 {
		return f64Prop(this.inner, "pointerId");
	}

	get pointer_type(): RuntimeString {
		return stringProp(this.inner, "pointerType");
	}

	get repeat(): Bool {
		return boolProp(this.inner, "repeat");
	}

	get screen_x(): F64 {
		return f64Prop(this.inner, "screenX");
	}

	get screen_y(): F64 {
		return f64Prop(this.inner, "screenY");
	}

	get shift_key(): Bool {
		return boolProp(this.inner, "shiftKey");
	}

	get target(): EventTarget {
		return new EventTarget(this.inner.target);
	}

	get time_stamp(): F64 {
		return new F64(this.inner.timeStamp);
	}

	prevent_default(): void {
		this.inner.preventDefault();
	}

	stop_immediate_propagation(): void {
		this.inner.stopImmediatePropagation();
	}

	stop_propagation(): void {
		this.inner.stopPropagation();
	}
}

export class EventTarget {
	constructor(private readonly inner: EventTargetLike) {}

	get checked(): Bool {
		return boolProp(this.inner, "checked");
	}

	get id(): RuntimeString {
		return stringProp(this.inner, "id");
	}

	get name(): RuntimeString {
		return stringProp(this.inner, "name");
	}

	get text_content(): RuntimeString {
		return new RuntimeString(
			this.inner instanceof Node ? (this.inner.textContent ?? "") : "",
		);
	}

	get value(): RuntimeString {
		return stringProp(this.inner, "value");
	}
}

function boolProp(source: unknown, name: string): Bool {
	return new Bool(
		typeof source === "object" && source !== null && name in source
			? Boolean((source as Record<string, unknown>)[name])
			: false,
	);
}

function f64Prop(source: unknown, name: string): F64 {
	const value =
		typeof source === "object" && source !== null && name in source
			? (source as Record<string, unknown>)[name]
			: 0;
	return new F64(
		typeof value === "number" && Number.isFinite(value) ? value : 0,
	);
}

function stringProp(source: unknown, name: string): RuntimeString {
	const value =
		typeof source === "object" && source !== null && name in source
			? (source as Record<string, unknown>)[name]
			: "";
	return new RuntimeString(value == null ? "" : globalThis.String(value));
}

File: crates/topcoat-runtime/src/surrogate/event.rs
//! Types that describe DOM events to runtime expressions. They exist so that
//! event handler bodies type-check in Rust. They are never constructed on
//! the server, and the browser reads their fields from the real DOM event.

use std::marker::PhantomData;

use crate::{BoolSurrogate, F64Surrogate, StringSurrogate};

/// The DOM event passed to an event handler closure.
///
/// Each field reads the DOM property named in its documentation. A property
/// the event does not have reads as `false`, `0.0`, or an empty string, and
/// a number that is not finite reads as `0.0`.
///
/// ```rust
/// # use topcoat::{Result, context::Cx, runtime::{Event, signal}, view::*};
/// # #[component]
/// # async fn example(cx: &Cx) -> Result<impl View> {
/// let text = signal(cx, String::new);
///
/// Ok(view! {
///     <input @input=$(|e: Event| text.set(e.target.value))>
/// })
/// # }
/// ```
pub struct Event {
    /// Whether the Alt key was held (`altKey`).
    pub alt_key: BoolSurrogate,
    /// Whether the event bubbles up through the DOM (`bubbles`).
    pub bubbles: BoolSurrogate,
    /// The mouse button that changed state (`button`).
    pub button: F64Surrogate,
    /// The mouse buttons held down, as a bit mask (`buttons`).
    pub buttons: F64Surrogate,
    /// Whether the event can be canceled (`cancelable`).
    pub cancelable: BoolSurrogate,
    /// The horizontal pointer position in the viewport (`clientX`).
    pub client_x: F64Surrogate,
    /// The vertical pointer position in the viewport (`clientY`).
    pub client_y: F64Surrogate,
    /// The physical key pressed (`code`).
    pub code: StringSurrogate,
    /// Whether the Control key was held (`ctrlKey`).
    pub ctrl_key: BoolSurrogate,
    /// The element the handler is attached to (`currentTarget`).
    pub current_target: EventTarget,
    /// The inserted characters of an input event (`data`).
    pub data: StringSurrogate,
    /// Whether the default action was prevented (`defaultPrevented`).
    pub default_prevented: BoolSurrogate,
    /// The horizontal scroll amount of a wheel event (`deltaX`).
    pub delta_x: F64Surrogate,
    /// The vertical scroll amount of a wheel event (`deltaY`).
    pub delta_y: F64Surrogate,
    /// The scroll amount on the z axis of a wheel event (`deltaZ`).
    pub delta_z: F64Surrogate,
    /// The event's name, such as `"click"` (`type`).
    pub event_type: StringSurrogate,
    /// The kind of change of an input event (`inputType`).
    pub input_type: StringSurrogate,
    /// Whether the event happened during text composition (`isComposing`).
    pub is_composing: BoolSurrogate,
    /// The key value of the key pressed (`key`).
    pub key: StringSurrogate,
    /// Whether the Meta key was held (`metaKey`).
    pub meta_key: BoolSurrogate,
    /// The horizontal pointer movement since the last event (`movementX`).
    pub movement_x: F64Surrogate,
    /// The vertical pointer movement since the last event (`movementY`).
    pub movement_y: F64Surrogate,
    /// The horizontal pointer position within the target (`offsetX`).
    pub offset_x: F64Surrogate,
    /// The vertical pointer position within the target (`offsetY`).
    pub offset_y: F64Surrogate,
    /// The horizontal pointer position in the document (`pageX`).
    pub page_x: F64Surrogate,
    /// The vertical pointer position in the document (`pageY`).
    pub page_y: F64Surrogate,
    /// The id of the pointer that caused the event (`pointerId`).
    pub pointer_id: F64Surrogate,
    /// The kind of pointer, such as `"mouse"` or `"touch"` (`pointerType`).
    pub pointer_type: StringSurrogate,
    /// Whether the key is held down and repeating (`repeat`).
    pub repeat: BoolSurrogate,
    /// The horizontal pointer position on the screen (`screenX`).
    pub screen_x: F64Surrogate,
    /// The vertical pointer position on the screen (`screenY`).
    pub screen_y: F64Surrogate,
    /// Whether the Shift key was held (`shiftKey`).
    pub shift_key: BoolSurrogate,
    /// The element the event was dispatched to (`target`).
    pub target: EventTarget,
    /// The time the event was created, in milliseconds (`timeStamp`).
    pub time_stamp: F64Surrogate,
    _priv: PhantomData<()>,
}

impl Event {
    /// Prevents the browser's default action for the event
    /// (`preventDefault()`).
    pub fn prevent_default(&self) {
        unreachable!();
    }

    /// Stops the event from reaching further elements
    /// (`stopPropagation()`).
    pub fn stop_propagation(&self) {
        unreachable!();
    }

    /// Stops the event from reaching further elements and other handlers on
    /// the same element (`stopImmediatePropagation()`).
    pub fn stop_immediate_propagation(&self) {
        unreachable!();
    }
}

/// The element an [`Event`] refers to.
///
/// Like on [`Event`], a property the element does not have reads as `false`
/// or an empty string.
pub struct EventTarget {
    /// Whether a checkbox or radio button is checked (`checked`).
    pub checked: BoolSurrogate,
    /// The element's `id`.
    pub id: StringSurrogate,
    /// The element's `name`.
    pub name: StringSurrogate,
    /// The text content of the element (`textContent`).
    pub text_content: StringSurrogate,
    /// The current value of a form control (`value`).
    pub value: StringSurrogate,
}
